src/internet_identity/src/vc_mvp.rs

```diff
@@ -0,0 +1,247 @@
+use crate::assets::CertifiedAssets;
+use crate::delegation::check_frontend_length;
+use crate::{delegation, hash, state, update_root_hash, LABEL_SIG, MINUTE_NS};
+use candid::Principal;
+use canister_sig_util::CanisterSigPublicKey;
+use ic_cdk::api::{data_certificate, time};
+use ic_cdk::trap;
+use ic_certified_map::{Hash, HashTree};
+use identity_core::common::Url;
+use identity_core::convert::FromJson;
+use identity_credential::credential::{Credential, CredentialBuilder, Subject};
+
+use canister_sig_util::signature_map::SignatureMap;
+use internet_identity_interface::internet_identity::types::vc_mvp::{
+    GetIdAliasResponse, IdAliasCredentials, PreparedIdAlias, SignedIdAlias,
+};
+use internet_identity_interface::internet_identity::types::{
+    AnchorNumber, FrontendHostname, IdentityNumber,
+};
+use serde::Serialize;
+use serde_bytes::ByteBuf;
+use serde_json::json;
+use vc_util::{
+    did_for_principal, vc_jwt_to_jws, vc_signing_input, vc_signing_input_hash, AliasTuple,
+    II_CREDENTIAL_URL_PREFIX, II_ISSUER_URL,
+};
+
+// The expiration used for signatures
+const SIGNATURE_EXPIRATION_PERIOD_NS: u64 = 10 * MINUTE_NS;
+
+pub struct InvolvedDapps {
+    pub(crate) relying_party: FrontendHostname,
+    pub(crate) issuer: FrontendHostname,
+}
+
+pub async fn prepare_id_alias(
+    identity_number: IdentityNumber,
+    dapps: InvolvedDapps,
+) -> PreparedIdAlias {
+    state::ensure_salt_set().await;
+    check_frontend_length(&dapps.relying_party);
+    check_frontend_length(&dapps.issuer);
+
+    let id_alias_principal = get_id_alias_principal(identity_number, &dapps);
+    let seed = calculate_id_alias_seed(identity_number, &dapps);
+    let canister_id = ic_cdk::id();
+    let canister_sig_pk = CanisterSigPublicKey::new(canister_id, seed.to_vec());
+
+    let rp_tuple = AliasTuple {
+        id_alias: id_alias_principal,
+        id_dapp: delegation::get_principal(identity_number, dapps.relying_party.clone()),
+    };
+    let issuer_tuple = AliasTuple {
+        id_alias: id_alias_principal,
+        id_dapp: delegation::get_principal(identity_number, dapps.issuer.clone()),
+    };
+
+    let (rp_id_alias_jwt, rp_msg_hash) = id_alias_jwt_and_msg_hash(&rp_tuple, &canister_sig_pk);
+    let (issuer_id_alias_jwt, issuer_msg_hash) =
+        id_alias_jwt_and_msg_hash(&issuer_tuple, &canister_sig_pk);
+
+    state::signature_map_mut(|sigs| {
+        add_signature(sigs, rp_msg_hash, seed);
+        add_signature(sigs, issuer_msg_hash, seed);
+    });
+    update_root_hash();
+    PreparedIdAlias {
+        canister_sig_pk_der: ByteBuf::from(canister_sig_pk.to_der()),
+        rp_id_alias_jwt,
+        issuer_id_alias_jwt,
+    }
+}
+
+fn id_alias_jwt_and_msg_hash(
+    alias_tuple: &AliasTuple,
+    canister_sig_pk: &CanisterSigPublicKey,
+) -> (String, Hash) {
+    let credential_jwt = prepare_id_alias_jwt(alias_tuple);
+    let signing_input =
```

Comment:
Is there ever a case where you don't do
`vc_signing_input_hash(vc_signing_input(jwt, pk))`?

Might be worth having a convenience function in the library for that?